docs/sphinx/heps/hep2.rst

```diff
@@ -0,0 +1,312 @@
+HEP 2: The public API for |hpx| 2.0.0
+=====================================
+
+Motivation
+----------
+
+|hpx| has evolved to have a rather ad-hoc organization of functionality into
+namespaces, headers, and libraries. This HEP specifies the public API, and
+through which namespaces and headers the public API is accessible, for |hpx|
+2.0.0.
+
+Implementation
+--------------
+
+The public API is presented as
+
+- ``header name``
+  - ``hpx::x``
+  - ``hpx::y``
+
+where including ``header name`` gives you access to at least ``hpx::x`` and
+``hpx::y``. Functionality may be organized differently internally (both headers
+and namespaces).
+
+Functionality that corresponds to, or extends, standard library functionality,
+will be in the same namespace as in the standard library with ``std`` replaced
+by ``hpx``, and a header ``abc`` replaced by ``hpx/abc.hpp``. Functionality that
+is proposed for the standard library, and may thus be in ``std::experimental``
+and a header ``experimental/abc`` does not need to be in the
+``hpx::experimental`` namespace or a ``hpx/experimental`` header if the
+functionality is considered stable in |hpx|. Distributed extensions to local
+parallelism and concurrency will be in the ``hpx::distributed`` namespace and
+``hpx/distributed`` headers.
+
+Utilities
+.........
+
+- ``hpx/functional.hpp``
+  - ``hpx::function``
+  - ``hpx::unique_function``
+  - ``hpx::function_nonser``
+  - ``hpx::unique_function_nonser``
+  - ``hpx::bind``
+  - ``hpx::bind_front``
+  - ``hpx::bind_back``
+  - ``hpx::invoke``
+  - ``hpx::invoke_fused``
+  - ``hpx::result_of``
+  - more?
+
+- ``hpx/tuple.hpp``
+  - ``hpx::tuple``
+
+- ``hpx/any.hpp``
+  - ``hpx::any``
+
+- ``hpx/optional.hpp``
+  - ``hpx::optional``
+
+- ``hpx/format.hpp``
+  - ``hpx::format``?
+  - other utilities that go with ``format``
+
+- ``hpx/chrono.hpp``
+  - ``hpx::high_resolution_clock``
+  - ``hpx::high_resolution_timer``
+  - ``hpx::scoped_timer``
+
+- ``hpx/exception.hpp``
+  - ``hpx::exception``
+  - ``HPX_THROW_EXCEPTION``
+
+- ``hpx/system_error.hpp``
+  - ``hpx::error_code``
+
+- ``hpx/assert.hpp``
+  - ``HPX_ASSERT``
+
+- ``hpx/serialization.hpp``
+  - what functionality actually needs to be public?
+  - Some classes have intrusive serialization in which case there's no need to
+    include ``serialization.hpp`` to get the serialization support; others have
+    it implemented in the serialization module. Should we just state that using
+    serialization for any class requires including ``hpx/serialization.hpp``
+    even though it may already be included transitively?
+
+Local concurrency and parallelism
+.................................
+
+- ``hpx/future.hpp``
+  - ``hpx::future``
+  - ``hpx::promise``
+  - ``hpx::shared_future``
+  - ``hpx::async``
+  - ``hpx::sync``
+  - ``hpx::apply``
+  - ``hpx::dataflow``
+  - ``hpx::launch_policy`` (or ``hpx::execution::launch_policy``?)
+  - ``hpx::is_future`` and friends
```

Comment:
I think we should keep our traits in `namespace traits`.

Reply:
A `traits` namespace never made sense to me.

Reply:
Something like this should at least not be in `hpx::traits`: https://en.cppreference.com/w/cpp/algorithm/is_execution_policy. I'm mostly indifferent about where non-std traits go, but I'd probably slightly prefer just `hpx` (or `hpx::distributed`).

Reply:
+0